akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala

```diff
@@ -444,44 +486,63 @@ import akka.util.Helpers.toRootLowerCase
     (fromVersion, manifestClassName)
   }
 
+  private lazy val lz4Factory = LZ4Factory.fastestInstance()
+  private lazy val lz4Compressor = lz4Factory.fastCompressor()
+  private lazy val lz4Decompressor = lz4Factory.safeDecompressor()
+
   def compress(bytes: Array[Byte]): Array[Byte] = {
-    compressionAlgorithm match {
-      case Compression.Off => bytes
-      case Compression.GZip(largerThan) =>
-        if (bytes.length > largerThan) compressGzip(bytes) else bytes
-    }
+    compress(ByteBuffer.wrap(bytes))
   }
 
-  private def compressGzip(bytes: Array[Byte]): Array[Byte] = {
-    val bos = new ByteArrayOutputStream(BufferSize)
-    val zip = new GZIPOutputStream(bos)
-    try zip.write(bytes)
-    finally zip.close()
-    bos.toByteArray
+  def compress(input: ByteBuffer): Array[Byte] = {
+    val bytes = input.array()
```

Comment:
There is too much juggling back and forth between ByteBuffer and Array[Byte] now. Could we keep this PR based on Array[Byte], and then add the full ByteBuffer support separately. It's ok to convert to ByteBuffer for lz4 if that simplifies things.